apollo-router/src/plugins/authentication.rs

```diff
@@ -0,0 +1,944 @@
+                            request.context,
+                            format!(
+                                "Header Value: '{jwt_value_untrimmed}' is not correctly formatted. prefix should be '{}'",
+                                my_config.header_value_prefix
+                            ),
+                            StatusCode::BAD_REQUEST,
+                        );
+                    }
+
+                    // Split our string in (at most 2) sections.
+                    let jwt_parts: Vec<&str> = jwt_value.splitn(2, ' ').collect();
+                    if jwt_parts.len() != 2 {
+                        return failure_message(
+                            request.context,
+                            format!("Header Value: '{jwt_value}' is not correctly formatted. Missing JWT"),
+                            StatusCode::BAD_REQUEST,
+                        );
+                    }
+
+                    // Trim off any trailing white space (not valid in BASE64 encoding)
+                    let jwt = jwt_parts[1].trim_end();
+
+                    // Try to create a valid header to work with
+                    let jwt_header = match decode_header(jwt) {
+                        Ok(h) => h,
+                        Err(e) => {
+                            return failure_message(
+                                request.context,
+                                format!("'{jwt}' is not a valid JWT header: {e}"),
+                                StatusCode::BAD_REQUEST,
+                            );
+                        }
+                    };
+
+                    // Try to find the kid of the header
+                    let kid = match jwt_header.kid {
+                        Some(k) => k,
+                        None => {
+                            return failure_message(
+                                request.context,
+                                "Missing kid value from JWT header".to_string(),
+                                StatusCode::BAD_REQUEST,
+                            );
+                        }
+                    };
+
+                    // Get the JWKS here
+                    let closure_jwks = my_jwks.clone();
+                    // If we ever find that:
+                    //  - we can't retrieve a JWKS (a)
+                    //  - the retrieved JWKS is None (b)
+                    //  - we have a kid that we don't know about (c)
+                    // We need to do some additional processing.  (I've tagged with comments a/b/c below)
+                    let err_cleanup = move |cache_clear| {
+                        // If cache_clear is set, then we clear the cache so that subsequent
+                        // requests will retrieve new JWKS.
+                        // The COOLDOWN controls attempts to retrieve based on a new "kid", but not
+                        // repeated attempts due to failure to retrieve JWKS.
+                        if cache_clear {
```

Comment:
why is `cache_clear` used here? Was there a previous version with an "else" branch? Because here `cache_clear` could be removed and we can replace `err_cleanup(true)` with `err_cleanup()`, and remove the calls to `err_cleanup(false)` at lines 363 (a) and 375 (b).
It also feels like `err_cleanup` might not be the right name for a rate limiter

Reply:
Good catch. You are right, there was a branch here originally. I'll clean this up as you suggest.

Reply:
done: 1bd8ed69f1d810adef32d3ad57d20d02328c0f6d